### DISCUSSION TITLE
Use readonly keyword for lambdas to enforce that captures are effectively readonly

### DISCUSSION THREAD

**Discussion by HaloFour** [dotnet/csharplang#5947] [Language Ideas]
There seems to be quite a bit of contention about adding [`readonly` locals/parameters to C#](https://github.com/dotnet/csharplang/issues/188).  There are concerns that repurposing the `readonly` keyword is too verbose and that local mutations are effectively safe and don't necessitate a language feature to prevent.  While I do think that intent is important I agree about the relatively safety in cases except for captures, where a local effectively becomes a field and may find its lifetime significantly extended and may be shared across any number of threads.

Recently the C# added a [new form of lambda](https://github.com/dotnet/csharplang/issues/275) to disable captures as an optimization.  By applying the `static` keyword to the lambda the compiler will prevent that lambda from capturing any identifiers from the enclosing scope.

What I would like to propose is to also support applying the `readonly` keyword to the lambda.  What this would do is to make any capture from the enclosing scope to be "effectively readonly".  That would mean that captured locals and parameters are no longer reassignable, not within the lambda and not within the enclosing method:

```cs
int x = 123;
var fn = readonly () => x + 1;  // compiler error
x = 456;
```

### Questions:
1. Should the capture only be readonly within the lambda?  If so, does that mean mutations from the outside scope are still observable, or would the capture be effectively a copy?
2. Would the compiler error occur on the attempt to reassign, or the capture within the lambda?

**Comment by jnm2** [comment]
I was expecting the compiler error to be on the lambda's usages of `x`. Similar to how static lambdas would be if you reference things restricted by that keyword.

**Reply by HaloFour** [reply]
I just checked Java and it does error on the capture, not the reassignment.  I'll change the example but leave the open question for the sake of discussion.

<img width="696" alt="image" src="https://user-images.githubusercontent.com/10102290/159742529-dfde30fd-9df4-4e05-9d1e-c0b0ceab75e0.png">

**Comment by 333fred** [comment]
Ok, this isn't what I assumed you meant in discord. I assumed that you meant the capture was readonly _in the lambda_, not everywhere. Not a fan of the everywhere spooky action at a distance thing in this.

**Reply by HaloFour** [reply]
This is a point I'm willing to discuss.  I was thinking of the implementation as being like Java, which forces the locals themselves to be readonly.

**Reply by 333fred** [reply]
I'm not a fan of a lambda declaration affecting the containing context's of x like this. I think it would be confusing and hard for users to reason about.

**Reply by HaloFour** [reply]
What I originally suggested is exactly how Java works and I don't think that behavior has been considered confusing or "spooky".

My concern with making the capture readonly only within the lambda is that it still leaves the problem of observable mutations from the enclosing method, or any other lambda capturing that same local.

**Reply by jnm2** [reply]
@333fred That concern is fixed by having the warnings appear on usages of the mutable locals inside the lambda, right?

**Reply by 333fred** [reply]
I'd have to think about it, but I suspect not.

**Reply by jnm2** [reply]
What if the language says all locals and parameters have a classification that says whether they are possibly mutated (though this tracking can be skipped where not used), and then the readonly lambda is the only location where anything is affected by the keyword?

**Comment by TahirAhmadov** [comment]
Did you mean to write `var fn = readonly () => x += 1;  // compiler error`?

**Reply by HaloFour** [reply]
No, my opinion is that because `x` is captured that `x` should be effectively readonly.  I don't want any observable mutation.

**Reply by TahirAhmadov** [reply]
How is returning `x + 1` a mutation though? Wouldn't that be a read-only operation on `x`?

**Reply by HaloFour** [reply]
Yes, the problem isn't with the lambda body itself, the problem is that the lambda captures `x` and there is a mutation/reassignment of `x` from outside of the lambda.  That's what I want to prevent.

**Reply by jnm2** [reply]
`x = 456` is a mutation that changes the value of `x + 1`. When this happens to a captured lambda variable, it's probably unintentional.

**Reply by jnm2** [reply]
Basically the value of `readonly` is: I no longer have to look for mutations of `x` *outside* the lambda and then think through all the ways in which the timing of the invocation(s) of the lambda might interact with those mutations.

**Reply by TahirAhmadov** [reply]
Oooooh, I see. 
Hmm it looks strange to me. I know what you are saying - Java does this and they are happy, but looking at it with an open mind, it looks off. I would much rather push for readonly locals.

**Reply by HaloFour** [reply]
@TahirAhmadov 

> I would much rather push for readonly locals.

As would I, but I'm hedging my bets on alternative options.

An argument against readonly locals is that local mutation is harmless due to the scope of the local.  My rebuttal is that captures expand the scope of locals to that of fields.

**Reply by TahirAhmadov** [reply]
I think perhaps this feature can be added alongside readonly locals? Basically, `readonly` lambda can only reference `readonly` (`let`) locals.

**Reply by jnm2** [reply]
I am skeptical of the value of readonly locals from personal experience with TypeScript which has `let` vs `const`. (Let's not discuss the reasons for this here; that belongs in #188.) Multiple LDM members also don't believe it has enough value. But I do like this proposal. I also liked an idea that came up on Discord (from ~~@HaloFour I think?~~ @YairHalberstadt) of a `byval` keyword for lambdas which would copy instead of capture.

**Reply by HaloFour** [reply]
@jnm2 

@YairHalberstadt brought up copying captures.  That's a feature of C++ lambdas.  It would work but I think it would also be more complicated and likely result in less efficient code generation as I'm not sure that'd be possible without allocating a separate display class at each lambda declaration to ensure that you have a copy of all of the captures.
